tests/testStatsdClient.cpp

```diff
@@ -1,65 +1,150 @@
+    std::vector<std::string> messages, expected;
+    std::thread server(mock, std::ref(mock_server), std::ref(messages));
+
+    // Set a new config that has the client send messages to a proper address that can be resolved
+    StatsdClient client("localhost", 8125, "sendRecv.", batchSize);
+    throwOnError(client);
+
+    // TODO: I forget if we need to wait for the server to be ready here before sending the first stats
+    //  is there a race condition where the client sending before the server binds would drop that clients message
+
+    for(int i = 0; i < 3; ++i) {
+        // Increment "coco"
+        client.increment("coco");
+        throwOnError(client);
+        expected.emplace_back("sendRecv.coco:1|c");
+
+        // Decrement "kiki"
+        client.decrement("kiki");
+        throwOnError(client);
+        expected.emplace_back("sendRecv.kiki:-1|c");
+
+        // Adjusts "toto" by +3
+        client.seed(19);  // this seed gets a hit on the first call
+        client.count("toto", 2, 0.1f);
+        throwOnError(client);
+        expected.emplace_back("sendRecv.toto:2|c|@0.10");
+
+        // Record a gauge "titi" to 3
+        client.gauge("titi", 3);
+        throwOnError(client);
+        expected.emplace_back("sendRecv.titi:3|g");
+
+        // Record a timing of 2ms for "myTiming"
+        client.seed(19);
+        client.timing("myTiming", 2, 0.1f);
+        throwOnError(client);
+        expected.emplace_back("sendRecv.myTiming:2|ms|@0.10");
+
+        // TODO: should sampling rates above 1 error or be capped?
+        // Send a metric explicitly
+        client.send("tutu", 4, "c", 2.0f);
+        throwOnError(client);
+        expected.emplace_back("sendRecv.tutu:4|c|@2.00");
     }
+    
+    // Signal the mock server we are done
+    client.send("DONE", 0, "DONE");
 
-    // Send a metric explicitly
-    client.send("tutu", 4, "c", 2.0f);
-    if (!client.errorMessage().empty()) {
-        std::cerr << client.errorMessage() << std::endl;
-        return EXIT_FAILURE;
+    // Wait for the server to stop
+    server.join();
+
+    // Make sure we get the exactly correct output
+    if(messages != expected) {
+        std::cerr << "Unexpected stats received by server, got:" << std::endl;
+        for(const auto& message : messages) {
+            std::cerr << message << std::endl;
+        }
+        std::cerr << std::endl << "But we expected:" << std::endl;
+        for(const auto& message : expected) {
+            std::cerr << message << std::endl;
+        }
+        throw std::runtime_error("Unexpected stats");
     }
+}
+
+
+int main(int argc, char** argv) {
+    // If any of these tests fail they throw an exception, not catching makes for a nonzero return code
+
+    // general things that should be errors
+    testErrorConditions();
+    // reconfiguring how you are sending
+    testReconfigure();
+    // no batching
+    testSendRecv(0);
+    // background batching
+    testSendRecv(4);
```

Comment:
these are the different sets of tests i came up with. i should probably enable code coverage to see what is and isnt covered at the moment, but i'm hoping its getting close to 100% (other than the stuff in the unit test code that arent covered because we arent actualy hitting failures in the tests haha)